### PR TITLE
screencopy: handle compositor not setting damage

### PR DESCRIPTION
**Pull request by ifreund**
Damage the full output if the compositor didn't submit damage but did
submit a buffer.

### CODE REVIEW

File: types/wlr_screencopy_v1.c [hunk 1]
```diff
@@ -48,9 +48,20 @@ static void screencopy_damage_accumulate(struct screencopy_damage *damage) {
...
-	pixman_region32_union(region, region, &output->pending.damage);
-	pixman_region32_intersect_rect(region, region, 0, 0, output->width,
-		output->height);
+	// If the compositor did not submit damage but did submit a buffer,
+	if (!(output->pending.committed & WLR_OUTPUT_STATE_DAMAGE) &&
+		(output->pending.committed & WLR_OUTPUT_STATE_BUFFER)) {
+		// damage everything
+		int width, height;
+		wlr_output_transformed_resolution(output, &width, &height);
+		pixman_region32_union_rect(region, region, 0, 0, width, height);
```
Comment (ifreund) [inline types/wlr_screencopy_v1.c:57]: I'm not sure if I chose the right pixman function here, they don't seem to have any docs...

File: types/wlr_screencopy_v1.c [hunk 2]
```diff
@@ -48,9 +48,20 @@ static void screencopy_damage_accumulate(struct screencopy_damage *damage) {
 		return;
 	}
 
-	pixman_region32_union(region, region, &output->pending.damage);
-	pixman_region32_intersect_rect(region, region, 0, 0, output->width,
-		output->height);
+	// If the compositor did not submit damage but did submit a buffer,
+	if (!(output->pending.committed & WLR_OUTPUT_STATE_DAMAGE) &&
+		(output->pending.committed & WLR_OUTPUT_STATE_BUFFER)) {
```
Comment (emersion) [inline types/wlr_screencopy_v1.c:53]: Style nit: use two tabs for `if` condition continuation lines, to tell them apart from the `if` body.

File: types/wlr_screencopy_v1.c (hunk 1 above)
Comment (emersion) [reply types/wlr_screencopy_v1.c:57]: Yes, this sounds like the right function to use.

File: types/wlr_screencopy_v1.c [hunk 3]
```diff
@@ -48,9 +48,20 @@ static void screencopy_damage_accumulate(struct screencopy_damage *damage) {
...
+		// damage everything
+		int width, height;
+		wlr_output_transformed_resolution(output, &width, &height);
+		pixman_region32_union_rect(region, region, 0, 0, width, height);
+	} else {
+		// otherwise copy over the current damage
+		pixman_region32_union(region, region, &output->pending.damage);
+		pixman_region32_intersect_rect(region, region, 0, 0, output->width,
+			output->height);
```
Comment (emersion) [inline types/wlr_screencopy_v1.c:62]: Looking at it now, it sounds like this breaks when using an output transform? We should use `wlr_output_transformed_resolution` like above, I think.

Can you fix that in a different commit (makes bisecting easier if we find bugs about it)?

File: types/wlr_screencopy_v1.c [hunk 4]
```diff
@@ -48,9 +48,20 @@ static void screencopy_damage_accumulate(struct screencopy_damage *damage) {
...
-	pixman_region32_intersect_rect(region, region, 0, 0, output->width,
-		output->height);
+	// If the compositor did not submit damage but did submit a buffer,
+	if (!(output->pending.committed & WLR_OUTPUT_STATE_DAMAGE) &&
+		(output->pending.committed & WLR_OUTPUT_STATE_BUFFER)) {
+		// damage everything
+		int width, height;
+		wlr_output_transformed_resolution(output, &width, &height);
+		pixman_region32_union_rect(region, region, 0, 0, width, height);
+	} else {
```
Comment (emersion) [inline types/wlr_screencopy_v1.c:58]: I know the previous code wasn't doing it, but we should check for `output->pending.committed & WLR_OUTPUT_STATE_DAMAGE` here. Otherwise we might be reading invalid state from `output->pending.damage`.

Can you fix that in a different commit (makes bisecting easier if we find bugs about it)?

**Comment by ifreund** [comment]
Added your suggestions in separate commits and retested to make sure my use-case still works (which it does). If the commits look good to you I'd say this is good to go.

**Review by emersion — APPROVED** [review]
LGTM, thanks!

File: types/wlr_screencopy_v1.c [hunk 5]
```diff
@@ -48,18 +48,18 @@ static void screencopy_damage_accumulate(struct screencopy_damage *damage) {
 		return;
 	}
 
+	int width, height;
+	wlr_output_transformed_resolution(output, &width, &height);
```
Comment (any1) [inline types/wlr_screencopy_v1.c:52]: This looks like it will break stuff.

Comment (ammen99) [reply types/wlr_screencopy_v1.c:52]: Ah indeed, damage is accumulated in transformed coordinates, but when it is committed, it is in buffer coordinates ...

Comment (emersion) [reply types/wlr_screencopy_v1.c:52]: Ah, interesting. I suppose we should just revert to `output->{width,height}` then?

I'll also create an issue about removing all of this mess.

Comment (any1) [reply types/wlr_screencopy_v1.c:52]: I haven't confirmed that it breaks anything. I just thought this looked suspicious because I've never run into an issue with damage reporting on transformed outputs before. I dunno, maybe it fixes bugs that I didn't know about.

Comment (emersion) [reply types/wlr_screencopy_v1.c:52]: I don't see any damage conversion in `wlr_output_set_damage` nor in `wlr_output_commit`. So I believe `wlr_output_state.damage` is in buffer-local coords. The comment in the header file even says so:

```c
struct wlr_output_state {
	pixman_region32_t damage; // output-buffer-local coordinates
```

In fact, Sway already transforms the damage manually before calling `wlr_output_set_damage`:

```c
	enum wl_output_transform transform =
		wlr_output_transform_invert(wlr_output->transform);
	wlr_region_transform(&frame_damage, &output->damage->current,
		transform, width, height);
	wlr_output_set_damage(wlr_output, &frame_damage);
	pixman_region32_fini(&frame_damage);
```

tl;dr we should use `output->{width,height}` here.

Comment (emersion) [reply types/wlr_screencopy_v1.c:52]: https://github.com/swaywm/wlroots/issues/2518

Comment (ifreund) [reply types/wlr_screencopy_v1.c:52]: #2519
